token-lending/sdk/src/state/reserve.rs

```diff
@@ -403,22 +452,31 @@ impl Reserve {
             settle_amount,
             repay_amount,
             withdraw_amount,
+            bonus_rate,
         })
     }
 
     /// Calculate protocol cut of liquidation bonus always at least 1 lamport
+    /// the bonus rate is always >=1 and includes both liquidator bonus and protocol fee.
+    /// the bonus rate has to be passed into this function because bonus calculations are dynamic
+    /// and can't be recalculated after liquidation.
     pub fn calculate_protocol_liquidation_fee(
         &self,
         amount_liquidated: u64,
+        bonus_rate: Decimal,
     ) -> Result<u64, ProgramError> {
-        let bonus_rate = Rate::from_percent(self.config.liquidation_bonus).try_add(Rate::one())?;
-        let amount_liquidated_wads = Decimal::from(amount_liquidated);
-
-        let bonus = amount_liquidated_wads.try_sub(amount_liquidated_wads.try_div(bonus_rate)?)?;
+        if bonus_rate
+            < Decimal::one().try_add(Decimal::from_percent(self.config.protocol_liquidation_fee))?
+        {
+            msg!("bonus rate is less than one + protocol liquidation fee");
+            return Err(LendingError::MathOverflow.into());
+        }
```

Comment:
in hindsight, this check feels kinda weird bc i don't want the program to halt and brick liquidations if this condition is somehow false

Reply:
probably better to do some kind of max instead